File: agents/issue-reconciler/api/fetchers/board.ts
import type { GitHubClient } from '../client.ts';
import { PROJECT_ID, REPO_NAME } from '../../config.ts';

export type BoardItem = {
  itemId: string;
  issueNumber: number;
  status: string | null;
};

const QUERY = `
  query BoardItems($projectId: ID!, $cursor: String) {
    node(id: $projectId) {
      ... on ProjectV2 {
        items(first: 100, after: $cursor) {
          pageInfo { hasNextPage endCursor }
          nodes {
            id
            status: fieldValueByName(name: "Status") {
              ... on ProjectV2ItemFieldSingleSelectValue { name }
            }
            content {
              __typename
              ... on Issue {
                number
                repository { name }
              }
            }
          }
        }
      }
    }
  }
`;

type RawResponse = {
  node: {
    items: {
      pageInfo: { hasNextPage: boolean; endCursor: string | null };
      nodes: Array<{
        id: string;
        status: { name: string } | null;
        content: {
          __typename: string;
          number?: number;
          repository?: { name: string };
        } | null;
      }>;
    };
  } | null;
};

/**
 * Fetches every board item, paginated, filtered to this repo's issues. The
 * board is user-level and may hold items from other repos - never touch
 * those (see plan "Board scope warning").
 */
export async function fetchBoardItems(client: GitHubClient): Promise<BoardItem[]> {
  const items: BoardItem[] = [];
  let cursor: string | null = null;

  do {
    const data: RawResponse = await client.query<RawResponse>(QUERY, {
      projectId: PROJECT_ID,
      cursor,
    });

    const page = data.node?.items;
    if (!page) break;

    for (const node of page.nodes) {
      if (node.content?.__typename !== 'Issue') continue;
      if (node.content.repository?.name !== REPO_NAME) continue;
      if (node.content.number === undefined) continue;

      items.push({
        itemId: node.id,
        issueNumber: node.content.number,
        status: node.status?.name ?? null,
      });
    }

    cursor = page.pageInfo.hasNextPage ? page.pageInfo.endCursor : null;
  } while (cursor);

  return items;
}

File: agents/issue-reconciler/issue_reconciler/api/fetchers/board.py
from __future__ import annotations

from typing import TypedDict

from issue_reconciler.api.client import GitHubClient
from issue_reconciler.config import PROJECT_ID, REPO_NAME

_QUERY = """
  query BoardItems($projectId: ID!, $cursor: String) {
    node(id: $projectId) {
      ... on ProjectV2 {
        items(first: 100, after: $cursor) {
          pageInfo { hasNextPage endCursor }
          nodes {
            id
            status: fieldValueByName(name: "Status") {
              ... on ProjectV2ItemFieldSingleSelectValue { name }
            }
            content {
              __typename
              ... on Issue {
                number
                repository { name }
              }
            }
          }
        }
      }
    }
  }
"""


class BoardItem(TypedDict):
    item_id: str
    issue_number: int
    status: str | None


def fetch_board_items(client: GitHubClient) -> list[BoardItem]:
    """Fetches every board item, paginated, filtered to this repo's issues.
    The board is user-level and may hold items from other repos - never
    touch those (see plan "Board scope warning").
    """
    items: list[BoardItem] = []
    cursor = None

    while True:
        data = client.query(_QUERY, {"projectId": PROJECT_ID, "cursor": cursor})
        page = (data.get("node") or {}).get("items")
        if not page:
            break

        for node in page["nodes"]:
            content = node.get("content") or {}
            if content.get("__typename") != "Issue":
                continue
            repository = content.get("repository") or {}
            if repository.get("name") != REPO_NAME:
                continue

            status = node.get("status")
            items.append(
                {
                    "item_id": node["id"],
                    "issue_number": content["number"],
                    "status": status["name"] if status else None,
                }
            )

        if page["pageInfo"]["hasNextPage"]:
            cursor = page["pageInfo"]["endCursor"]
        else:
            break

    return items
